### ISSUE TITLE
SIGSEGV error: a large tensor cannot be created; how to reduce memory consumption of MKL-DNN.

### ISSUE THREAD

**Issue by mozga-intel** [oneapi-src/oneDNN#848]
Hi, 

When running our pre-model (this is a simple imitation of the LSTM model). While a test, I want to create a large LRN tensor, for example: (20758,500). I can see that ~ 170GB of memory is allocated for scratchpad computations. At first glance, I noticed that use_global_scratchpad is always true. Well, as a result, for a different oneDNN version, I got the error messages: as follows:
- v.1.3
-```Segmentation fault: 11```
- v.1.7
 -```mxnet.base.MXNetError: MXNetError: could not create a primitive```

For brevity: This error is visible for a large LSTM tensor. Yet, I debugged the code step-by-step and I think this is some sort of out of memory issue because if we shrink the input (first dim of input) then there will not be a **SIGSEGV** fault (we want to create tensor: ```void *ptr = malloc(size, platform::get_cache_line_size())``` Furthermore, if I change the shape from **(20758,500) to (10758,500)** then it will not segfault on this machine  But this threshold is different with different memory sizes. During execution, I noticed that the total size of the primitive allocated for a given instance of a tensor is equal: **187516940848 bytes ~ 170 GB**. As I mentioned before the following error is not visible for its naive implementation.  Furthermore, I tried to enable/disable external/internal cache mechanism to reduce the size of memory needed to create the following tensor: But I noticed, this error is still visible and large tensor cannot be created correctly.

1. Should it be possible to decrease memory consumption? 
2. Why do we need to allocate a lot of memory for a given tensor? 
3. How we can reduce the total size of memory which is needed to create a huge tensor? 

Thanks!.

**Comment by mozga-intel** [comment]
For brevity: 
For a given tensor: (input: 20758 , 500), we should allocate: ```500 ^ 2 * 20758 * 4 * 4 ~ aprox: 70 GB```.  Then, ```init_scratchpad(scratchpad_sz)``` function is called with a special variable ```scratchpad_sz```. The size of iscratchpad_sz is at least ```60782687168 bytes ~ 56 GB```. This function was debugged step-by-step. I got these results: 

1. There is a ```scratchpad.template book<float>(key_rnn_space, scratchpad_sz, 4096); 
https://github.com/oneapi-src/oneDNN/blob/rls-v1.6/src/cpu/rnn/ref_rnn.hpp#L221
2. .There is a **book** function called with a few params: (key, nelems, perf_align)
https://github.com/oneapi-src/oneDNN/blob/rls-v1.6/src/common/memory_tracking.hpp#L432
where: nelems 60782687168 - this is LSTM scratchpad and the key is equal 68.
3. Then, we have to register the book:
```registry_.book(make_key(prefix_, key), nelems * data_size, data_align, perf_align);```



nelems * data_size = 60782687168 * 4  = 243130748672 bytes = **230** GB`
```
(gdb) p nelems * sizeof(T)
$14 = 243130748672
(gdb)
````

**Comment by mozga-intel** [comment]
Additionally: DNNL_VERBOSE
* **[FAIL]** For a given tensor: **15758,500**
```
platform linux -- Python 3.7.3, pytest-4.3.1, py-1.8.0, pluggy-0.9.0 -- /home/mozga/anaconda3/bin/python
cachedir: .pytest_cache
rootdir: /home/mozga/incubator-mxnet/tests/python/unittest, inifile:
plugins: remotedata-0.3.1, openfiles-0.3.2, doctestplus-0.3.0, arraydiff-0.3
collecting ...
[[ 71499.42    77235.8     93174.87   ...  21490.842   63583.973
   16263.251 ]
 [ 46329.227   94072.664  122519.48   ... 115694.336   65904.27
   92317.56  ]
 [ 40436.414   36188.355   48598.98   ...  42781.21    66134.76
  104574.73  ]
 ...
 [ 23043.238   95707.305   51101.57   ...  50720.805   53281.203
   44237.617 ]
 [ 64195.676   79581.22    64110.11   ...   6855.8794  89362.58
   96547.266 ]
 [ 27026.588   30118.682    6783.8604 ...  95621.9     83589.016
   45634.883 ]]
<NDArray 15758x500 @cpu(0)>
dnnl_verbose,info,oneDNN v1.6.4 (commit 570ab9c2ea0915b048279438aebb063b8769149c)
dnnl_verbose,info,cpu,runtime:OpenMP
dnnl_verbose,info,cpu,isa:Intel AVX-512 with AVX512BW, AVX512VL, and AVX512DQ extensions
dnnl_verbose,info,gpu,runtime:none
dnnl_verbose,create:cache_miss,cpu,rnn,ref:any,forward_inference,src_layer_f32::blocked:abc:f0 src_iter_f32::blocked:abcd:f0 wei_layer_f32::blocked:abcde:f0 wei_iter_f32::rnn_packed::f0 bias_f32::blocked:abcd:f0 dst_layer_f32::blocked:abc:f0 dst_iter_f32::blocked:abcd:f0,,alg:vanilla_lstm direction:bidirectional_concat activation:undef,l1t15758mb500sic30slc100dhc30dlc60,0.723145
collected 0 items / 1 errors

====================================================================================== ERRORS =======================================================================================
___________________________________________________________________________ ERROR collecting test_lstm.py ___________________________________________________________________________
test_lstm.py:24: in <module>
    nd.waitall()
../../../python/mxnet/ndarray/ndarray.py:211: in waitall
    check_call(_LIB.MXNDArrayWaitAll())
../../../python/mxnet/base.py:246: in check_call
    raise get_last_ffi_error()
E   mxnet.base.MXNetError: MXNetError: could not create a primitive
````
* **[OK]** For a given tensor: (14758,500)
```
[[ 71499.42    77235.8     93174.87   ...  21490.842   63583.973
   16263.251 ]
 [ 46329.227   94072.664  122519.48   ... 115694.336   65904.27
   92317.56  ]
 [ 40436.414   36188.355   48598.98   ...  42781.21    66134.76
  104574.73  ]
 ...
 [ 15900.922   71976.04    89555.125  ...  31668.174   87506.78
   67947.61  ]
 [ 77669.67    41547.863  127347.01   ...  56141.6    123388.19
   22187.418 ]
 [ 41001.527  103037.78    43460.277  ...   3886.1426  84107.85
   65980.91  ]]
<NDArray 14758x500 @cpu(0)>
dnnl_verbose,info,oneDNN v1.6.4 (commit 570ab9c2ea0915b048279438aebb063b8769149c)
dnnl_verbose,info,cpu,runtime:OpenMP
dnnl_verbose,info,cpu,isa:Intel AVX-512 with AVX512BW, AVX512VL, and AVX512DQ extensions
dnnl_verbose,info,gpu,runtime:none
dnnl_verbose,create:cache_miss,cpu,rnn,ref:any,forward_inference,src_layer_f32::blocked:abc:f0 src_iter_f32::blocked:abcd:f0 wei_layer_f32::blocked:abcde:f0 wei_iter_f32::rnn_packed::f0 bias_f32::blocked:abcd:f0 dst_layer_f32::blocked:abc:f0 dst_iter_f32::blocked:abcd:f0,,alg:vanilla_lstm direction:bidirectional_concat activation:undef,l1t14758mb500sic30slc100dhc30dlc60,0.733887
dnnl_verbose,create:cache_miss,cpu,concat,simple:any,undef,src_f32::blocked:abdec:f0 src_f32::blocked:abdec:f0 dst_f32::blocked:abdec:f0,,axis:1,1x1x100x4x30:1x1x100x4x30 1x2x100x4x30,0.0549316
dnnl_verbose,exec,cpu,concat,simple:any,undef,src_f32::blocked:abdec:f0 src_f32::blocked:abdec:f0 dst_f32::blocked:abdec:f0,,axis:1,1x1x100x4x30:1x1x100x4x30 1x2x100x4x30,0.146973
dnnl_verbose,create:cache_miss,cpu,concat,simple:any,undef,src_f32::blocked:abdec:f0 src_f32::blocked:abdec:f0 dst_f32::blocked:abdec:f0,,axis:1,1x1x30x4x30:1x1x30x4x30 1x2x30x4x30,0.0498047
dnnl_verbose,exec,cpu,concat,simple:any,undef,src_f32::blocked:abdec:f0 src_f32::blocked:abdec:f0 dst_f32::blocked:abdec:f0,,axis:1,1x1x30x4x30:1x1x30x4x30 1x2x30x4x30,0.0429688
dnnl_verbose,create:cache_miss,cpu,reorder,jit:uni,undef,src_f32::blocked:abdec:f0 dst_f32::blocked:abcde:f0,,,1x2x100x4x30,0.184082
dnnl_verbose,create:cache_miss,cpu,reorder,rnn_weights_reorder,undef,src_f32::blocked:abdec:f0 dst_f32::rnn_packed::f0,,,1x2x30x4x30,0.0969238
dnnl_verbose,exec,cpu,reorder,jit:uni,undef,src_f32::blocked:abdec:f0 dst_f32::blocked:abcde:f0,,,1x2x100x4x30,0.0720215
dnnl_verbose,exec,cpu,reorder,rnn_weights_reorder,undef,src_f32::blocked:abdec:f0 dst_f32::rnn_packed::f0,,,1x2x30x4x30,0.158936
dnnl_verbose,exec,cpu,rnn,ref:any,forward_inference,src_layer_f32::blocked:abc:f0 src_iter_f32::blocked:abcd:f0 wei_layer_f32::blocked:abcde:f0 wei_iter_f32::rnn_packed::f0 bias_f32::blocked:abcd:f0 dst_layer_f32::blocked:abc:f0 dst_iter_f32::blocked:abcd:f0,,alg:vanilla_lstm direction:bidirectional_concat activation:undef,l1t14758mb500sic30slc100dhc30dlc60,29406.1
dnnl_verbose,create:cache_miss,cpu,inner_product,gemm:jit,forward_inference,src_f32::blocked:ab:f0 wei_f32::blocked:ab:f0 bia_f32::blocked:a:f0 dst_f32::blocked:ab:f0,,,mb14758ic30000oc3,0.219971
dnnl_verbose,exec,cpu,inner_product,gemm:jit,forward_inference,src_f32::blocked:ab:f0 wei_f32::blocked:ab:f0 bia_f32::blocked:a:f0 dst_f32::blocked:ab:f0,,,mb14758ic30000oc3,305.955

[[-5.2847597e-04 -3.4048291e-05 -1.2547804e-04]
 [-2.8190056e-03 -1.8457485e-03 -1.7332821e-03]
 [-5.0275103e-04 -3.7277718e-03 -1.5350026e-03]
 ...
 [ 8.7051926e-04  1.3417397e-03  5.4993277e-04]
 [ 1.5870768e-03  1.5777488e-03 -4.7814039e-05]
 [ 2.9817026e-03  1.1125103e-03 -4.3313904e-04]]
<NDArray 14758x3 @cpu(0)>
collected 0 items
```

**Comment by vpirogov** [comment]
Thank you for the report, @mozga-intel. Scratchpad is a temporary space used by primitive implemenation to improve performance. It may depend on the problem dimensions and the number of threads.

The difference in behavior between v1.3 and v1.7 is likely the result of improved handling for out of memory situations in the latter version. oneDNN v1.7 identifies out of memory condition and returns an error, while v1.3 segfaults.

@piotrchmiel, @dsciebu, could you please comment on details of LRN implementation with respect to scratchpad use?

**Comment by vpirogov** [comment]
I see that the report is actually about LSTM primitive, not LRN. The considerations related to scratchpad and out of memory conditions still apply though.

**Comment by emfomenk** [comment]
It was a typo. @mozga-intel meant LSTM.
There is also a suspicious place in the code (kudos to @mozga-intel for the investigation):
https://github.com/oneapi-src/oneDNN/blob/570ab9c2ea0915b048279438aebb063b8769149c/src/cpu/rnn/ref_rnn.hpp#L221
It might happen that `<float>` should be replaced with `<uint8_t>` there, in which case the requirements for scratchpad will dramatically decrease.

**Comment by vpirogov** [comment]
Adding @mgouicem.

**Comment by mgouicem** [comment]
@emfomenk  is definitely right,  the <float> should be changed to <uint8_t> when booking the scratchpad.
Currently, the memory consumption increases with the problem size for LSTM and RNN in general, as we keep intermediate states in the scratchpad (in particular it grows with the number of layers and timesteps).
If you are using forward_training, there is not much we can do since we need to keep intermediate states for the backward computation in the workspace.
If you are using forward_inference, the scratchpad should be smaller. If it is still too big for some applications, we can tweak the primitive behavior to reduce memory consumption (for example by less aggressively fusing matrix multiplications internally).

**Comment by vpirogov** [comment]
The fix is backported to rls-v1.6 and will be available in v1.6.5 patch release.
